snapshotter/metrics_integ_test.go

```diff
@@ -0,0 +1,165 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/firecracker-microvm/firecracker-containerd/firecracker-control/client"
+	"github.com/firecracker-microvm/firecracker-containerd/internal/integtest"
+	"github.com/firecracker-microvm/firecracker-containerd/proto"
+	"github.com/firecracker-microvm/firecracker-containerd/snapshotter/internal/integtest/stargz/fs/source"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	serviceDiscoveryEndpoint = "http://localhost:8080"
+)
+
+func TestSnapshotterMetrics_Isolated(t *testing.T) {
+	integtest.Prepare(t, integtest.WithDefaultNetwork())
+
+	vmID := 0
+
+	testTimeout := 30 * time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
+	defer cancel()
+
+	ctx = namespaces.WithNamespace(ctx, strconv.Itoa(vmID))
+
+	fcClient, err := integtest.NewFCControlClient(integtest.ContainerdSockPath)
+	defer fcClient.StopVM(ctx, &proto.StopVMRequest{VMID: strconv.Itoa(vmID)})
+	require.NoError(t, err, "Failed to create fccontrol client")
+
+	pullImageWithRemoteSnapshotterInVM(ctx, t, strconv.Itoa(vmID), fcClient)
+	verifyMetricsResponse(t, 1)
+}
+
+func TestSnapshotterMetricsMultipleVMs_Isolated(t *testing.T) {
+	integtest.Prepare(t, integtest.WithDefaultNetwork())
+
+	testTimeout := 300 * time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
+	defer cancel()
+
+	var wg sync.WaitGroup
+
+	numberOfVms := integtest.NumberOfVms
+	fcClient, err := integtest.NewFCControlClient(integtest.ContainerdSockPath)
+	require.NoError(t, err, "Failed to create fccontrol client")
+	for vmID := 0; vmID < numberOfVms; vmID++ {
+		defer func(ctxInner context.Context, id int) {
+			fcClient.StopVM(ctxInner, &proto.StopVMRequest{VMID: strconv.Itoa(id)})
+		}(namespaces.WithNamespace(ctx, strconv.Itoa(vmID)), vmID)
+
+		wg.Add(1)
+		go func(ctxInner context.Context, id int) {
+			defer wg.Done()
+			pullImageWithRemoteSnapshotterInVM(ctxInner, t, strconv.Itoa(id), fcClient)
+		}(namespaces.WithNamespace(ctx, strconv.Itoa(vmID)), vmID)
+
+	}
```

Comment:
It is better to not pass t to different goroutines.

https://pkg.go.dev/testing#T.FailNow

> FailNow must be called from the goroutine running the test or benchmark function, not from other goroutines created during the test. Calling FailNow does not stop those other goroutines. 

And testify's [require](https://github.com/stretchr/testify#require-package) functions call FailNow

Reply:
Ok, I'll attempt using an error channel rather than passing testing.T to these goroutines. Looks like we have the same potential issue here as well: https://github.com/firecracker-microvm/firecracker-containerd/blob/47ceb625562fadcd7d615f519a24086a45cf73cc/snapshotter/service_integ_test.go#L74-L81

Reply:
Returning an error from launchContainerWithRemoteSnapshotterInVM and using errgroup may be the way to go :)